`move` recommendation with closure and invalid lifetime is inserted at the wrong place

### Code

```Rust
#[derive(Default)]
struct A {}

struct B<'a> {
    a: &'a mut A,
}

struct C<'a, 'b> {
    b: &'b mut B<'a>,
}

fn some_fn<'a>(b: &mut B<'a>, f: impl FnOnce(&C<'a, '_>)) {}

fn main() {
    let some_closure = |_fs| {};

    for _ in [1] {
        let mut a = Default::default();
        let mut b = B { a: &mut a };
        some_fn(&mut b, |c| {
            some_closure(c);
        });
    }
}
```


### Current output

```
error[E0373]: closure may outlive the current block, but it borrows `a`, which is owned by the current block
  --> src/main.rs:19:28
   |
19 |         let mut b = B { a: &mut a };
   |                            ^^^^^^
   |                            |
   |                            `a` is borrowed here
   |                            may outlive borrowed value `a`
   |
note: block requires argument type to outlive `'1`
  --> src/main.rs:21:13
   |
21 |             some_closure(c);
   |             ^^^^^^^^^^^^
help: to force the closure to take ownership of `a` (and any other referenced variables), use the `move` keyword
   |
19 |         let mut b = B { a: move &mut a };
   |                            ++++
```


### Desired output

`move` should not be recommended as it does not solve any issue. This is the case even if put at the closure definition since the closure does not capture anything.


### Rationale and extra context

The recommendation for the `move` is wrong.

### Other cases

_No response_

### Anything else?

[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=1f00fb4be87086286aab7e8e3b52424d).

Rust version:

```
Nightly version: 1.72.0-nightly

(2023-06-21 065a1f5df9c2f1d93269)
```

--- a/compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs
+++ b/compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs
@@ -1730,18 +1730,19 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
             (
                 Some(name),
                 BorrowExplanation::UsedLater(LaterUseKind::ClosureCapture, var_or_use_span, _),
-            ) => self.report_escaping_closure_capture(
-                borrow_spans,
-                borrow_span,
-                &RegionName {
-                    name: self.synthesize_region_name(),
-                    source: RegionNameSource::Static,
-                },
-                ConstraintCategory::CallArgument(None),
-                var_or_use_span,
-                &format!("`{}`", name),
-                "block",
-            ),
+            ) if borrow_spans.for_generator() || borrow_spans.for_closure() => self
+                .report_escaping_closure_capture(
+                    borrow_spans,
+                    borrow_span,
+                    &RegionName {
+                        name: self.synthesize_region_name(),
+                        source: RegionNameSource::Static,
+                    },
+                    ConstraintCategory::CallArgument(None),
+                    var_or_use_span,
+                    &format!("`{}`", name),
+                    "block",
+                ),
             (
                 Some(name),
                 BorrowExplanation::MustBeValidFor {
@@ -1754,7 +1755,7 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                     span,
                     ..
                 },
-            ) if borrow_spans.for_generator() | borrow_spans.for_closure() => self
+            ) if borrow_spans.for_generator() || borrow_spans.for_closure() => self
                 .report_escaping_closure_capture(
                     borrow_spans,
                     borrow_span,

--- a/tests/ui/closures/issue-113087.rs
+++ b/tests/ui/closures/issue-113087.rs
@@ -1,0 +1,11 @@
+fn some_fn<'a>(_: &'a i32, _: impl FnOnce(&'a i32)) {}
+
+fn main() {
+    let some_closure = |_| {};
+
+    for a in [1] {
+        some_fn(&a, |c| { //~ ERROR does not live long enough
+            some_closure(c);
+        });
+    }
+}

--- a/tests/ui/closures/issue-113087.stderr
+++ b/tests/ui/closures/issue-113087.stderr
@@ -1,0 +1,16 @@
+error[E0597]: `a` does not live long enough
+  --> $DIR/issue-113087.rs:7:17
+   |
+LL |     for a in [1] {
+   |         - binding `a` declared here
+LL |         some_fn(&a, |c| {
+   |                 ^^ borrowed value does not live long enough
+LL |             some_closure(c);
+   |             ------------ borrow later captured here by closure
+LL |         });
+LL |     }
+   |     - `a` dropped here while still borrowed
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0597`.

